_studio/mfx_lib/vpp/src/mfx_vpp_hw.cpp

```diff
@@ -2962,6 +2962,28 @@ mfxStatus VideoVPPHW::Close()
 } // mfxStatus VideoVPPHW::Close()
 
 
+bool VideoVPPHW::UseCopyPassThrough(const DdiTask *pTask) const
+{
+    if (!m_config.m_bCopyPassThroughEnable
+        || IsRoiDifferent(pTask->input.pSurf, pTask->output.pSurf))
+    {
+        return false;
+    }
+
+    if (m_pCore->GetVAType() != MFX_HW_VAAPI || m_ioMode != D3D_TO_D3D)
+    {
+        return true;
+    }
+
+    // Under VAAPI for D3D_TO_D3D either CM copy or VPP pipeline are preferred by
+    // performance reasons. So, before enabling CopyPassThrough for a task we check
+    // for CM availability to do fallback to VPP if there is no CM.
+    // This can't be done in ConfigureExecuteParams() because CmCopy status is set later.
+    const VAAPIVideoCORE *vaapiCore = dynamic_cast<VAAPIVideoCORE*>(m_pCore);
+    assert(vaapiCore);
```

Comment:
I think here you can use following macro to simplify code + use assert + use traces:
```
MFX_CHECK_WITH_ASSERT(vaapiCore, false);

return vaapiCore->CmCopy();
```

Reply:
Yep.